ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala

```diff
@@ -0,0 +1,279 @@
+    * @param inputState: The input DAML state entries. Some state entries declared as input can be missing.
+    * @return The log entry to be committed and the DAML state updates.
+    */
+  def processSubmission(
+      engine: Engine,
+      config: Configuration,
+      entryId: DamlLogEntryId,
+      recordTime: Timestamp,
+      submission: DamlSubmission,
+      inputLogEntries: Map[DamlLogEntryId, DamlLogEntry],
+      inputState: Map[DamlStateKey, Option[DamlStateValue]]
+  ): (DamlLogEntry, Map[DamlStateKey, DamlStateValue]) = {
+
+    submission.getPayloadCase match {
+      case DamlSubmission.PayloadCase.ARCHIVE =>
+        val archive = submission.getArchive
+        val key = DamlStateKey.newBuilder.setPackageId(archive.getHash).build
+        (
+          DamlLogEntry.newBuilder
+            .setRecordTime(buildTimestamp(recordTime))
+            .setArchive(archive)
+            .build,
+          Map(
+            key -> DamlStateValue.newBuilder.setArchiveEntry(entryId).build
+          )
+        )
+      case DamlSubmission.PayloadCase.CONFIGURATION_ENTRY =>
+        (
+          DamlLogEntry.newBuilder
+            .setRecordTime(buildTimestamp(recordTime))
+            .setConfigurationEntry(submission.getConfigurationEntry)
+            .build,
+          Map.empty
+        )
+      case DamlSubmission.PayloadCase.TRANSACTION_ENTRY =>
+        processTransactionSubmission(
+          engine,
+          config,
+          entryId,
+          recordTime,
+          submission.getTransactionEntry,
+          inputLogEntries,
+          inputState
+        )
+
+      case DamlSubmission.PayloadCase.PAYLOAD_NOT_SET =>
+        throw new RuntimeException("DamlSubmission payload not set!")
+    }
+  }
+
+  private def processTransactionSubmission(
+      engine: Engine,
+      config: Configuration,
+      entryId: DamlLogEntryId,
+      recordTime: Timestamp,
+      txEntry: DamlTransactionEntry,
+      inputLogEntries: Map[DamlLogEntryId, DamlLogEntry],
+      inputState: Map[DamlStateKey, Option[DamlStateValue]]
+  ): (DamlLogEntry, Map[DamlStateKey, DamlStateValue]) = {
+
+    val txLet = parseTimestamp(txEntry.getLedgerEffectiveTime)
+
+    // 1. Verify that this is not a duplicate command submission.
+    val dedupKey = commandDedupKey(txEntry.getSubmitterInfo)
+    val dedupEntry = inputState(dedupKey)
+    val dedupCheckResult = dedupEntry.isEmpty
+
+    // 2. Verify that the ledger effective time falls within time bounds
+    val letCheckResult = config.timeModel.checkLet(
+      currentTime = recordTime.toInstant,
+      givenLedgerEffectiveTime = txLet.toInstant,
+      givenMaximumRecordTime =
+        parseTimestamp(txEntry.getSubmitterInfo.getMaximumRecordTime).toInstant
+    )
+
+    // 3. Verify that the input contracts are active.
+    // FIXME(JM): Does this need to be done separately, or is it enough to
+    // have lookupContract to check these?
+    val activenessCheckResult = true
```

Comment:
I believe it is enough. A written out justification will help becoming sure; and support readers in their reasoning.